Katib v1beta1 documentation update (#2312)
kubeflow/website: content/en/docs/components/katib/overview.md

+++
title = "Introduction to Katib"
description = "Overview of Katib for hyperparameter tuning and neural architecture search"
weight = 10
                    
+++

{{% beta-status
  feedbacklink="https://github.com/kubeflow/katib/issues" %}}

Katib is a Kubernetes-native project for automated machine learning (AutoML) —
it's a system for hyperparameter tuning and neural architecture search.
Katib supports a number of machine learning frameworks, including
TensorFlow, MXNet, PyTorch, XGBoost, and others.
Learn more about AutoML at [fast.ai](https://www.fast.ai/2018/07/16/auto-ml2/),
[Google Cloud](https://cloud.google.com/automl),
[Microsoft Azure](https://docs.microsoft.com/en-us/azure/machine-learning/concept-automated-ml#automl-in-azure-machine-learning) or
[Amazon SageMaker](https://aws.amazon.com/blogs/aws/amazon-sagemaker-autopilot-fully-managed-automatic-machine-learning/).

> Katib supports a lot of various AutoML algorithms, such as
> Bayesian optimization, Tree of Parzen Estimators, Random Search,
> Covariance Matrix Adaptation Evolution Strategy, Hyperband,
> Neural Architecture Search based on ENAS, Differentiable Architecture Search and
> many more. Further algorithms are under development.

Reviewer: In the entire paragraph, I'd recommend adding links to arXiv and other original posts that describe each AutoML algorithm in detail (you did that in another guide that listed the algorithms). Also, maybe it's best to spell out ENAS? Just a suggestion.

```suggestion
Katib supports a lot of various AutoML algorithms, such as
Bayesian optimization, Tree of Parzen Estimators, Random Search,
Covariance Matrix Adaptation Evolution Strategy, Hyperband,
Neural Architecture Search based on ENAS, Differentiable Architecture Search and
many more. Additional algorithm support is coming soon.
```

Reviewer: Was Google's NAS with RL paper the earliest research published on AutoML, btw? 🤔 "Neural Architecture Search with Reinforcement Learning" - https://arxiv.org/abs/1611.01578

Author: Yes, it was one of the earliest.
@8bitmp3 Do we want to add it also ?
Our ENAS implementation is based on this paper: https://github.com/kubeflow/katib/tree/master/pkg/suggestion/v1beta1/nas/enas#about-the-efficient-neural-architecture-search.

Reviewer: It was just a suggestion - thought it could be relevant if you found it some use. Up to you

The [Katib project](https://github.com/kubeflow/katib) is open source.
The [developer guide](https://github.com/kubeflow/katib/blob/master/docs/developer-guide.md)
is a good starting point for developers who want to contribute to the project.

> This page introduces the concepts of hyperparameter tuning, neural

Reviewer: ```suggestion
This guide introduces the concepts of hyperparameter tuning, neural
```

> This page introduces the concepts of hyperparameter tuning, neural

Reviewer: Maybe it makes sense to move this sentence towards the beginning, since every guide starts with: "Dear reader, this doc does XYZ. XYZ is a product that does ABC.."

> architecture search, and the Katib system as a component of Kubeflow.

Reviewer: It may make sense to define "NAS" early in the guide before referring to it as "NAS"

                    
+++

{{% beta-status
  feedbacklink="https://github.com/kubeflow/katib/issues" %}}

Katib is a Kubernetes-native project for automated machine learning (AutoML).
For example, to run hyperparameter tuning or neural architecture search.
Learn more about AutoML at [fast.ai](https://www.fast.ai/2018/07/16/auto-ml2/),
[Google Cloud](https://cloud.google.com/automl),
[Microsoft Azure](https://docs.microsoft.com/en-us/azure/machine-learning/concept-automated-ml#automl-in-azure-machine-learning) or
[Amazon SageMaker](https://aws.amazon.com/blogs/aws/amazon-sagemaker-autopilot-fully-managed-automatic-machine-learning/).

Katib is the framework agnostic to machine learning frameworks.
It can tune hyperparameters of applications written in any language
of the users' choice and natively supports many machine learning frameworks,
such as TensorFlow, MXNet, PyTorch, XGBoost, and others.

The [Katib project](https://github.com/kubeflow/katib) is open source.
The [developer guide](https://github.com/kubeflow/katib/blob/master/docs/developer-guide.md)

> is a good starting point for developers who want to contribute to the project.

Reviewer: Can you please use my previous suggestion (old text from the existing Markdown) and throw in the cloud links, so we don't reinvent the wheel? There's quite a lot of review in this PR :)

Author: @8bitmp3 Sure, I've made section as you said:
https://github.com/kubeflow/website/blame/b946be02fb367f8e9c2fb5d88a7818ddd7c18ee8/content/en/docs/components/katib/overview.md#L11-L31

                    
+++

{{% beta-status
  feedbacklink="https://github.com/kubeflow/katib/issues" %}}

> Katib is Kubernetes-native project for automated machine learning (AutoML).
> Use Katib for automated tuning of your machine learning (ML) model's hyperparameters and architecture.

Reviewer: Suggestion:

```suggestion
Katib is a Kubernetes-native project for automated machine learning (AutoML).
You can use Katib for automated tuning of your machine learning (ML) model's hyperparameters and architecture.
```

This page introduces the concepts of hyperparameter tuning, neural
architecture search, and the Katib system as a component of Kubeflow.

## Hyperparameters and hyperparameter tuning

_Hyperparameters_ are the variables that control the model training process.

> For example:

Reviewer: ```suggestion
They include:
```

> - Learning rate.
> - Number of layers in a neural network.
> - Number of nodes in each layer.

Reviewer: ```suggestion
- The learning rate.
- The number of layers in a neural network.
- The number of nodes in each layer.
```

Hyperparameter values are not _learned_. In other words, in contrast to the
node weights and other training _parameters_, the model training process does
not adjust the hyperparameter values.

_Hyperparameter tuning_ is the process of optimizing the hyperparameter values
to maximize the predictive accuracy of the model. If you don't use Katib or a
similar system for hyperparameter tuning, you need to run many training jobs
yourself, manually adjusting the hyperparameters to find the optimal values.

Automated hyperparameter tuning works by optimizing a target variable,
also called the _objective metric_, that you specify in the configuration for
the hyperparameter tuning job. A common metric is the model's accuracy
in the validation pass of the training job (_validation-accuracy_). You also
specify whether you want the hyperparameter tuning job to _maximize_ or
_minimize_ the metric.

For example, the following graph from Katib shows the level of validation accuracy

> for various combinations of hyperparameter values (learning rate, number of
> layers, and optimizer):

Reviewer: ```suggestion
for various combinations of hyperparameter values (the learning rate, the number of
layers, and the optimizer):
```

## The Katib project

Katib is a Kubernetes-based system for hyperparameter tuning and neural
architecture search. Katib supports a number of ML frameworks, including
TensorFlow, MXNet, PyTorch, XGBoost, and others.

The [Katib project](https://github.com/kubeflow/katib) is open source.
The [developer guide](https://github.com/kubeflow/katib/blob/master/docs/developer-guide.md)

> is a good starting point for developers who want to contribute to the project.

Reviewer: I think lines 87-95 about "the Katib project" should be right in the beginning of `overview.md` (see earlier comment).

## Katib interfaces

You can use the following interfaces to interact with Katib:

- A web UI that you can use to submit experiments and to monitor your results.
  See the [getting-started
  guide](/docs/components/katib/hyperparameter/#katib-ui)
  for information on how to access the UI.
  The Katib home page within Kubeflow looks like this:

  <img src="/docs/images/katib/katib-home.png"
      alt="The Katib home page within the Kubeflow UI"
      class="mt-3 mb-3 border border-info rounded">

- A gRPC API. See the [API reference](/docs/reference/katib/v1beta1/katib).

- Command-line interfaces (CLIs):

>   - **Kfctl** is the Kubeflow CLI that you can use to install and configure

Reviewer: ```suggestion
  - **kfctl** is the Kubeflow CLI that you can use to install and configure
```

(do we use it with a capitalized letter K? Not 100% sure)

Author: Agree, changed.
